functions/database/database.js

```diff
@@ -0,0 +1,316 @@
+    const releaseData = {
+      state: RELEASE_STATES.SCHEDULED,
+      releaseName: release.releaseName,
+      releaseOperator: release.releaseOperator,
+      codeFreezeDate: release.codeFreezeDate,
+      releaseDate: release.releaseDate,
+      snapshotBranchName: snapshotBranchName,
+      snapshotBranchLink: snapshotBranchLink,
+      releaseBranchName: releaseBranchName,
+      releaseBranchLink: releaseBranchLink,
+      isComplete: false,
+      buildArtifactStatus: "",
+      buildArtifactConclusion: "",
+      buildArtifactLink: "",
+      buildArtifactJobId: "",
+    };
+    batch.set(newReleaseRef, releaseData);
+  });
+
+  await batch.commit();
+}
+
+/**
+   * Delete the releases in Firestore with code freeze dates that are in the
+   * future. This is intented to be used when scheduling new upcoming releases,
+   * but we need to delete the old upcoming releases before.
+   */
+async function deleteUpcomingReleases() {
+  const today = Timestamp.now();
+  const releasesRef = db.collection("releases");
+  const query = releasesRef.where("codeFreezeDate", ">", today);
+  const querySnapshot = await query.get();
+  const batch = db.batch();
+
+  querySnapshot.forEach((doc) => {
+    batch.delete(doc.ref);
+  });
+
+  await batch.commit();
+}
+
+/**
+   * Update a release in Firestore.
+   *
+   * @param {string} releaseId The ID of the release to update.
+   * @param {string} releaseData The new data to set for the release.
+   * @throws {Error} If the release state update fails.
+   */
+async function updateRelease(releaseId, releaseData) {
+  validateRelease(releaseData);
+  const releaseDoc = db.collection("releases").doc(releaseId);
+  await releaseDoc.update(releaseData);
+}
+
+/**
+ * Updates the state of a release in the database.
+ *
+ * @param {string} releaseId The ID of the release to update
+ * @param {string} newState The new state to set for the release
+ * @throws {Error} If the release state update fails, or if the
+ * release state is invalid.
+ */
+async function updateReleaseState(releaseId, newState) {
+  // Verify that the state is one of the valid states
+  if (!Object.values(RELEASE_STATES).includes(newState)) {
+    throw new Error(`Invalid release state: ${newState}`);
+  }
+
+  const releaseDoc = db.collection("releases").doc(releaseId);
+  await releaseDoc.update({state: newState});
+}
+
+/**
+ * Creates new library release documents for each version in the libraryVersions
+ * object, and deletes any existing library versions associated with the
+ * release.
+ *
+ * @param {Object} libraryVersions Object mapping library names to their
+ * versions.
+ * @param {string} releaseId The ID of the associated release.
+ */
+async function updateLibrariesForRelease(libraryVersions, releaseId) {
+  const batch = db.batch();
+
+  // Delete all previous library versions for this release
+  const previousLibrariesSnapshot = await db.collection("libraries")
```

Comment:
tbh I'm not sure how I feel about this. I think it should deff be in the same batch for atomic reasons, but it makes me wonder if there should be a design choice to pass around batches (which doesn't sound nice anyways)

Reply:
I added some functions that accept batches to perform batched writes here https://github.com/firebase/firebase-release-dashboard/pull/8/commits/030399c9ed011642c29c1d0d1c946d65c8e3d17f.